include/xsimd/math/xsimd_scalar.hpp

```diff
@@ -453,6 +481,58 @@ namespace xsimd
         return tmp * tmp;
     }
 
+    template<typename T, class = typename std::enable_if<std::is_scalar<T>::value>::type>
+    T sadd(const T& lhs, const T& rhs)
+    {
+        if (std::numeric_limits<T>::is_signed)
+        {
+            if ((lhs > 0) && (rhs > std::numeric_limits<T>::max() - lhs))
+            {
+                return std::numeric_limits<T>::max();
+            }
+            else if ((lhs < 0) && (rhs < std::numeric_limits<T>::lowest() - lhs))
+            {
+                return std::numeric_limits<T>::lowest();
+            }
+            else {
+                return lhs + rhs;
+            }
+        }
+        else
+        {
+            if (rhs > std::numeric_limits<T>::max() - lhs)
+            {
+                return std::numeric_limits<T>::max();
+            }
+            else
+            {
+                return lhs + rhs;
+            }
+
+        }
+    }
+
+    template<typename T, class = typename std::enable_if<std::is_scalar<T>::value>::type>
+    T ssub(const T& lhs, const T& rhs)
+    {
+        if (std::numeric_limits<T>::is_signed)
+        {
+            return sadd(lhs, (T)-rhs);
+        }
+        else
+        {
+            if (lhs < rhs)
+            {
+                return std::numeric_limits<T>::lowest();
+            }
+            else
+            {
+                return lhs - rhs;
+            }
+
+        }
+    }
+
```

Comment:
Do you think it would make sense to have `sadd` and `ssub` in `xsimd_fallback.hpp` and have `xsimd_scalar.hpp` depending on it? The idea is to have the `math` folder depenting on the `types` folder and to avoid cyclic dependencies between folders. Even if it's not a big deal at the folder level, it helps avoiding circular include at the header level.

If there are many functions like this (or more to come), it could even make sense to create an implmeentation file in `types` on which both `xsimd_fallback.hpp` and `xsimd_scalar.hpp` depend.

Reply:
> If there are many functions like this (or more to come), it could even make sense to create an implementation file in types on which both xsimd_fallback.hpp and xsimd_scalar.hpp depend.

I second that. We already have `ipow`, `sadd`, `ssub` ... which name do you suggest?

Reply:
xsimd/types/xsimd_math_utils.hpp or xsimd/types/xsimd_basic_math.hpp or xsimd/types/xsimd_common_math.hpp ?